### ISSUE TITLE
Having Trouble with Clustered Cloudburst

### ISSUE THREAD

**Issue by kraftp** [hydro-project/cloudburst#63]
Hello!

I've been trying to get Cloudburst working in cluster mode to reproduce some of its results (particularly Figures 11-12 from the VLDB paper) and I'm having some trouble.  I'm running all experiments on AWS in us-east-1 on an Ubuntu 20.04 instance.  I created a Hydro cluster successfully using the suggested command:

    python3 -m hydro.cluster.create_cluster -m 1 -r 1 -f 1 -s 1

I then tried to connect to the cluster with Cloudburst.  Installing the Cloudburst Python requirements didn't work (Python 3.8.10, Pip 20.0.2)--the Pandas installation failed and the required PyArrow version (0.14.1) didn't exist.  To fix this, I used the most recent versions of Pandas and PyArrow, which both installed successfully.  I installed all other dependencies (including Anna) successfully using the provided scripts. I then tried to connect to a cluster and perform a simple operation using the suggested commands:

```python3
>>> elb_address = "a91cbd914819441efb3414f96e794c2a-891365270.us-east-1.elb.amazonaws.com" # the function service IP from the hydro cluster
>>> my_ip = "172.31.70.53" # the internal IP address of my EC2 instance
>>> from cloudburst.client.client import CloudburstConnection
>>> cloudburst = CloudburstConnection(elb_address , my_ip , local=False)
>>> incr = lambda _, a: a + 1
>>> cloud_incr = cloudburst.register(incr, 'incr')
>>> cloud_incr(1).get()
```

The first few commands all work, but the final command hangs.  Here's a stack trace of where it's hanging:
  
    File "<stdin>", line 1, in <module>
    File "/home/ubuntu/cloudburst/cloudburst/shared/future.py", line 23, in get
      obj = self.kvs_client.get(self.obj_id)[self.obj_id]
    File "/usr/local/lib/python3.8/dist-packages/anna/client.py", line 85, in get
      worker_addresses[key] = (self._get_worker_address(key))
    File "/usr/local/lib/python3.8/dist-packages/anna/client.py", line 247, in _get_worker_address
      addresses = self._query_routing(key, port)
    File "/usr/local/lib/python3.8/dist-packages/anna/client.py", line 278, in _query_routing
      response = recv_response([key_request.request_id],
    File "/usr/local/lib/python3.8/dist-packages/anna/zmq_util.py", line 27, in recv_response
      resp = rcv_sock.recv()
    File "zmq/backend/cython/socket.pyx", line 788, in zmq.backend.cython.socket.Socket.recv
    File "zmq/backend/cython/socket.pyx", line 824, in zmq.backend.cython.socket.Socket.recv
    File "zmq/backend/cython/socket.pyx", line 186, in zmq.backend.cython.socket._recv_copy
    File "zmq/backend/cython/checkrc.pxd", line 12, in zmq.backend.cython.checkrc._check_rc

Do you know what's causing the problem and how to fix it?  Thank you!

**Comment by vsreekanti** [comment]
Hi Peter, thanks for asking. 

tl;dr: I believe why this is hanging is because the EC2 instance from which you are running the code doesn't have the correct ports open. The ports you need to open can be found [here](https://github.com/hydro-project/anna/blob/master/client/python/anna/common.py#L16-L17). You also will need to use the **public** address of your EC2 instance rather than the private address. 

**Explanation about opening ports**: Cloudburst by default stores all results in Anna, so `.get()` is essentially syntax candy for doing a `GET` operation on Anna for a UUID generated by Cloduburst. Anna is an asynchronous-only system, so when you make a request to Anna, it doesn't operate in a request-response mode, but you send it an address/port pair where you will listen for the results. The ports I linked above are the ports used to get responses from Anna, so you will need to open those in order to receive responses.

You can do this by going to your EC2 console and modifying the security group for the appropriate EC2 instance to allow requests on the ports in the link above. 

**Explanation about public vs. private IP**: Private IPs on EC2 are only accessible within the VPC that you are running in. When you create a Kubernetes cluster with the cluster creation scripts, you are creating the cluster in a separate VPC. You could peer the VPCs to each other, which will allow IP addresses to resolve across them, but the easier thing to do is to use the instance's public IP address. (This is also why you need to open ports on the security group -- when you use a public instance, you can only access ports that are explicitly opened on EC2.)

**Comment by kraftp** [comment]
Thanks!  Opening the ports and fixing the IP address let me receive a reply from Anna.  Unfortunately, I have a new problem,  the reply is always:

     ("ERROR: 'bytes' object is not callable", b'\x18\x03')

For reference, my script is:

```python3
from cloudburst.client.client import CloudburstConnection
cloudburst = CloudburstConnection("a2e694d5bfc9a408aa1021e59c324cc3-1110458798.us-east-1.elb.amazonaws.com" , "3.220.174.131" , local=False)
incr = lambda _, a: a + 1
cloud_incr = cloudburst.register(incr, 'incr')
result = cloud_incr(1).get()
print("The result is:", result)
```

**Comment by kraftp** [comment]
Alternatively @vsreekanti if it would be easier, do you have a working image/AMI of the Cloudburst client I could use to connect to a Hydro cluster and run benchmarks?

**Comment by vsreekanti** [comment]
Hey @kraftp, sorry to hear you're still struggling with this. My guess is that the error message around the bytes object not being callable might have to do with a Python version mismatch or a serialization issue on the client-side. A couple things that you could try doing:
1. Looking at the k8s logs for the function executors pods -- if you run `kubectl get pods`, you should see something with the title `function-*`, and if you run `kubectl logs <pod-name>`, that might give you a more detailed idea about what is happening. 
2. If that doesn't help, you could add some logging statements [here](https://github.com/hydro-project/cloudburst/blob/master/cloudburst/server/executor/call.py) to see where the error might be happening. To make sure the cluster is using your updated code, you need to modify the YAML configs that you can find [here](https://github.com/hydro-project/cluster/blob/master/hydro/cluster/yaml/ds/function-ds.yml#L58-L61).
3. You could try using one of the Cloudburst or benchmark Docker images, which you can find [here](https://hub.docker.com/u/hydroproject). 

Unfortunately, I don't have an AMI handy that I can share -- the team that was working on Cloudburst has graduated and moved on to other things, so no one is actively maintaining the project.

Hope this is helpful!

**Comment by kraftp** [comment]
Thanks!  Using the Docker image got things working--like you said, this was almost certainly a Python dependency/compatibility issue.

**Comment by vsreekanti** [comment]
Glad that fixed it! We're not making any strong assumptions about Python versioning in the code, so you could probably update the system by modifying the scheduler/executor Dockerfiles to use a newer Python version.
